validator/src/main/java/com/amazon/aoc/validators/XrayValidator.java

```diff
@@ -64,63 +54,11 @@ public void init(
     this.xrayService = new XRayService(context.getRegion());
     this.caller = caller;
     this.context = context;
-  }
-
-  @Override
-  public void validate() throws Exception {
-    // 2 retries for calling the sample app to handle the Lambda case,
-    // where first request might be a cold start and have an additional unexpected subsegment
-    int successes = 0;
-    while (successes < 5) {
-      // Call sample app and get locally stored trace
-      String traceId = this.getAppTraceId();
-      List<String> traceIdList = Collections.singletonList(traceId);
-      ;
-      log.info("value of Sample App traceId: {}", traceId);
-
-      // Retry 5 times to since segments might not be immediately available in X-Ray service
-      RetryHelper.retry(
-          5,
-          Integer.parseInt(GenericConstants.SLEEP_IN_MILLISECONDS.getVal()),
-          true,
-          () -> {
-            // get retrieved trace from x-ray service
-            Map<String, Object> retrievedTrace = this.getRetrievedTrace(traceIdList);
-            log.info("value of retrieved trace map: {}", retrievedTrace);
-            Set<String> set =
-                retrievedTrace.keySet().stream()
-                    .filter(s -> s.endsWith("collector-id"))
-                    .collect(Collectors.toSet());
-            if (set.size() < 2) {
-              log.error("not enough spans in trace map (need at least 2)");
-              log.info("==========================================");
-              throw new BaseException(ExceptionCode.NOT_ENOUGH_SPANS);
-            }
-            String targetCollectorId = null;
-            for (String collectorId : set) {
-              if (targetCollectorId == null) {
-                targetCollectorId = retrievedTrace.get(collectorId).toString();
-              } else {
-                if (!targetCollectorId.equals(retrievedTrace.get(collectorId).toString())) {
-                  log.info("id values do not match");
-                  log.info("value of target id: {}", targetCollectorId);
-                  log.info("value of retrieved id: {}", retrievedTrace.get(collectorId).toString());
-                  log.info("==========================================");
-                  throw new BaseException(ExceptionCode.COLLECTOR_ID_NOT_MATCHED);
-                }
-              }
-            }
-          });
-
-      successes += 1;
-      log.info("Total number of successful runs: {}", successes);
-    }
-
-    log.info("validation is passed for path {}", caller.getCallingPath());
+    this.expectedTrace = expectedTrace;
   }
 
   // this method will hit get trace from x-ray service and get retrieved trace
-  private Map<String, Object> getRetrievedTrace(List<String> traceIdList) throws Exception {
+  public Map<String, Object> getRetrievedTrace(List<String> traceIdList) throws Exception {
```

Comment:
This whole method throws me off a bit. It is called `getRetrievedTrace` (singular) but takes in a list. It then proceeds to use the xrayService to query for a list of traces by id. Then it just blindly grabs all the segments from the first trace.

Shouldn't this method take in a single traceId? 

Is `xrayService.listTraceByIds()` used somewhere else where a List<Trace> data structure is actually needed?

Reply:
I think the weird ergonomics of this method force us to write not great code in the classes that extend this abstract class. IMO, it would be cleaner to have two methods `getRetrivedTrace(traceId)` and a `getRetrievedTraces(List<String> traceIdList).

Reply:
Nit: I don't love the word `retrieved` here. Personally, it took me some time to get use to it when diving into the Validator. The validator is basically a big testing application. It does assertions against actual versus expected trace ids and data fields. I think `actual` would be more suiting....but that's just like my opinion man.

Reply:
I believe it was a list of traceIds due to this [method](https://docs.aws.amazon.com/xray/latest/api/API_BatchGetTraces.html) which is the only way to get segments from a trace and only takes in a list of traceIds.  For readability, I have added a method to our XrayService to get the segments from a single traceId and so I can add another method to our XrayValidator called `getActualTrace()` which is less confusing and prevents us from being constricted to providing a list input.